### PR TITLE
Added geometric distribution to scipy stats

### LINKED ISSUE #3191: Require for Geometric Distribution
It seems that `jax` does not have geometric distribution in its random module like `scipy.stats`. Will it be added? I really need it.

### PR DESCRIPTION
**Pull request by gsp-27**
Fixes #3191

Hi, I have made an attempt to add ```geometric distribution```.

As this is my first attempt at contribution, I apologize if I am missing something.
If I am indeed missing something please let me know. I will try to fix it.

I have done a few tests to check the code, please let me know if there is test-suite format that I should follow to allow for better coverage of possible scenarios

Thanks

### CODE REVIEW

**Comment by gsp-27** [comment]
@googlebot I fixed it.

**Comment by gsp-27** [comment]
@googlebot  I fixed it.

**Comment by jakevdp** [comment]
Thanks! Looks great so far. There still seems to be an issue with the CLA; also, please add a test case modeled after the other distribution tests in this file: https://github.com/google/jax/blob/master/tests/scipy_stats_test.py

**Comment by hawkinsp** [comment]
@gsp-27 It seems that the CLA bot is unhappy with the email address `gspat@Rishis-MacBook-Pro.local`. That seems like an email address you didn't intend to use anyway, given it is clearly not a public address. You should amending the author information of your existing commits and push a version without that email address in it.

(And in general, this probably means you need to fix the Git author information config on your Macbook...)

**Comment by gsp-27** [comment]
Hi,
Thanks for the advice. I will fix it and push a commit without that address. Also I will try to add the tests.

**Comment by jakevdp** [comment]
Hi – I don't think it's enough to add a new commit with the updated address: you need to amend the existing commits with your updated address.

**Comment by gsp-27** [comment]
Hi,
I am trying to do ```git rebase -i 0eace80a6e12ded1ad7ecb3e7d6ccb41b84e97e4``` and then fixing the author of commits by doing edit in the opened window, but somehow it still is not reflecting.

Do you have any suggestions how could this be fixed?

I am sorry for the hassle and junk history I created in this process.

**Comment by hawkinsp** [comment]
Take a look at https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit

It's not sufficient to change in the text in the `git rebase -i` editor, you actually need to edit the individual commits.

**Comment by jakevdp** [comment]
At this point there are so many empty commits on your branch, I'd just recommend starting over. Do a soft reset to the commit before your branch started, add and re-commit your changes, and then force push (note, if you've done any rebasing that is not reflected here, you'll need to modify the commit hash that you reset to)
```
$ git reset --soft 0eace80
$ git add . -u
$ git commit -m "Added geometric distribution to scipy stats"
$ git push -f origin master
```

**Comment by gsp-27** [comment]
Thanks so much for the suggestion and sorry for the bother.

**Comment by jakevdp** [comment]
Awesome! CLA bot is appeased :grinning: Next we'll need a test of the new distribution here: https://github.com/google/jax/blob/master/tests/scipy_stats_test.py

**Comment by gsp-27** [comment]
Hi,

I have added the tests

File: jax/scipy/stats/geom.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2018 Google LLC
```
Comment (jakevdp) [inline jax/scipy/stats/geom.py:1]: 2020

File: jax/scipy/stats/geom.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from ..special import xlogy, xlog1py
+
+@_wraps(osp_stats.geom.logpmf, update_doc=False)
+def logpmf(k, p, loc=0):
+    k, p, loc = jnp._promote_args_inexact("geom.logpmf", k, p, loc)
+    zero = jnp._constant_like(k, 0)
+    one = jnp._constant_like(k, 1)
+    x = lax.sub(k, loc)
+    log_probs = xlog1py(lax.sub(x, one), -p) + lax.log(p)
+    return jnp.where(lax.le(x, zero), -jnp.inf, log_probs)
```
Comment (jakevdp) [inline jax/scipy/stats/geom.py:29]: I think you might need the "double where" trick to avoid breaking the gradient at zero:

```python
invalid = lax.le(x, zero)
x = jnp.where(invalid, 1, x) 
log_probs = xlog1py(lax.sub(x, one), -p) + lax.log(p)
return jnp.where(invalid, -jnp.inf, log_probs)
```

it would be worth checking, though.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great! Thanks for the contribution

File: jax/scipy/stats/geom.py (hunk 2 above)
Comment (gsp-27) [reply jax/scipy/stats/geom.py:29]: Hi,
I might be interpreting something wrong here. Please correct me if I am wrong. I tried the following things

First with no change in code
```
g = lsp_stats.geom.logpmf
grad_g = jax.grad(g)
print(grad_g(0.0, 0.5)
```

and then same thing with change suggested above
for first case I got, ```DeviceArray(-0., dtype=float32)``` and for second I got ```DeviceArray(0, dtype=float32)```

Comment (gsp-27) [reply jax/scipy/stats/geom.py:29]: is this what we wanted to check?

Comment (jakevdp) [reply jax/scipy/stats/geom.py:29]: Yes, thanks – it appears that the change is not necessary then.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Running some quick internal tests before merging.

**Comment by jakevdp** [comment]
Thanks!

**Comment by gsp-27** [comment]
Thanks :) Is there anything else I can try?

**Comment by jakevdp** [comment]
Take a look at #70 for more functionality to be implemented

**Comment by gsp-27** [comment]
okay thanks :)
